File: a2/receiver.py
# Parameters:
#   $1 - hostname for the network emulator
#   $2 - UDP port number used by the link emulator to receive ACKs from the receiver
#   $3 - UDP port number used by the receiver to receive data from the emulator
#   $4 - name of the file into which the received data is written

import sys
import threading
import time
from socket import *
from packet import packet

T_ACK = 0
T_DATA = 1
T_EOT = 2

class Receiver:

  def __init__(self):
    self.packet_num = 0

  def run(self):
    global T_DATA
    global T_EOT
    global server_socket
    global emu_addr
    global emu_port
    global f_output
    global f_arrival

    while True:
      data, server_addr = server_socket.recvfrom( 2048 )
      p = packet.parse_udp_data(data)

      f_arrival.write(str(p.seq_num) + "\n")

      send_packet = None
      if p.type == T_DATA:
        if p.seq_num == self.packet_num % packet.SEQ_NUM_MODULO:          
          f_output.write(p.data) # read and write to file
          self.packet_num += 1
        send_packet = packet.create_ack(self.packet_num - 1)
        if not self.packet_num == 0: # if received atleast one good packet
          server_socket.sendto(send_packet.get_udp_data(), (emu_addr, emu_port))
      elif p.type == T_EOT:
        send_packet = packet.create_eot(self.packet_num)
        server_socket.sendto(send_packet.get_udp_data(), (emu_addr, emu_port))
        break

############# MAIN STARTS HERE #############

def check_args(args):
  if len(args) != 4:
    raise Exception
  try:
    emu_addr = args[0]
    emu_port = int(args[1])
    receiver_port = int(args[2])
    file_name = args[3]
  except Exception as e:
    raise e

  return emu_addr, emu_port, receiver_port, file_name

# Check if command-line arguments are valid
try:
  emu_addr, emu_port, receiver_port, file_name = check_args(sys.argv[1:])
except Exception:
  print("ERROR: invalid number of arguments")
  exit(-1)

# Create UDP Socket
server_socket = socket(AF_INET, SOCK_DGRAM) 
server_socket.bind(('', receiver_port))

# Generate three log files
f_output = open(file_name,"w")
f_arrival = open("arrival.log","w")

# Start receiving
receiver = Receiver()
receiver.run()

# Close socket
server_socket.close()

# Close files
f_output.close()
f_arrival.close()

File: a2/src/receiver.java
import java.io.*;
import java.net.*;

public class receiver {
    private static final int ACK = 0;
	private static final int DATA = 1;
	private static final int EOT = 2;

    private static InetAddress emulatorAddr;
    private static int emulatorPort;
    private static int receiverPort;
    private static String fileName;

    private PrintWriter outputLog, arrivalLog;

    DatagramSocket socket;

    private int nextSeqNum = 0;

    public static void main(String[] args) throws Exception{
        receiver r = new receiver();

        // init fields
        r.init(args);

        // start receiver
        r.start();

        // close files and sockets
        r.close();
    }

    private void init(String[] args) throws Exception {
        if (args.length != 4) {
            System.err.println("Receiver: invalid number of arguments");
            System.exit(-1);
        }

        emulatorAddr = InetAddress.getByName(args[0]);
        try {
            // validate port numbers
            emulatorPort = Integer.parseInt(args[1]);
            receiverPort = Integer.parseInt(args[2]);
        } catch (NumberFormatException e) {
            System.err.println("Receiver: invalid port number given");
            System.exit(-1);
        }
        fileName = args[3];

        // create socket
        try {
            socket = new DatagramSocket(receiverPort);
        } catch (SocketException e) {
            System.err.println("Receiver: given port is not available");
        }

        // create files
        try {
            outputLog = new PrintWriter(new BufferedWriter(new FileWriter(fileName)));
            arrivalLog = new PrintWriter(new BufferedWriter(new FileWriter("arrival.log")));
        } catch (IOException e) {
            System.err.println("Sender: failed to create files");
        }
    }

    private void close() throws Exception {
        // close files
        try {
            outputLog.close();
            arrivalLog.close();
        } catch (Exception e) {
            System.err.println("Receiver: cannot close files");
        }

        // close socket
        try {
            socket.close();
        } catch (Exception e) {
            System.err.println("Receiver: failed to close socket");
        }
    }

    private void start() throws Exception {
        while(true) {
            byte[] data = new byte[512];
            DatagramPacket receivePacket = new DatagramPacket(data, data.length);

            socket.receive(receivePacket);

            packet dataPacket = packet.parseUDPdata(receivePacket.getData());

            if (dataPacket.getType() == EOT) {
                // send eot packet
                send( packet.createEOT(dataPacket.getSeqNum()) );
                break;
            } else if (dataPacket.getType() == DATA) {
                arrivalLog.println(dataPacket.getSeqNum());

                if (dataPacket.getSeqNum() == nextSeqNum) {
                    // send acked packet
                    send( packet.createACK(dataPacket.getSeqNum()) );

                    // write data to file
                    byte[] bytes = dataPacket.getData();
                    outputLog.print( new String(bytes) );

                    nextSeqNum = (nextSeqNum + 1) % 32;
                } else {
                    if (nextSeqNum == 0) continue;
                    // send ack for most recent packet
                    send( packet.createACK(nextSeqNum - 1) );
                }
            }
        }
    }

    private void send(packet p) throws Exception {
        byte[] data = p.getUDPdata();
        DatagramPacket sendPacket = new DatagramPacket(data, data.length, emulatorAddr, emulatorPort);
        socket.send(sendPacket);
    }
}
